lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+      }
+    }
+  }
+}
+
+void ExecutionState::insertIntoNodeCtx(const DAGNode *node, Placeholder *P,
+                                       Tensor &&T) {
+  // Get a raw pointer to the input Context for the node. It should have
+  // been created in the constructor.
+  auto ctxIt = inputCtxs_.find(node);
+
+  if (ctxIt == inputCtxs_.end()) {
+    assert(false && "Input context not found but should exist!");
+  }
+
+  Context *ctx = (ctxIt->second).get();
+  assert(ctx && "Input context for node is null");
+
+  // Insert the placeholder-tensor pair.
+  std::lock_guard<std::mutex> lock(contextMtx_);
+  ctx->insert(P, std::move(T));
+}
+
+std::unique_ptr<Context>
+ExecutionState::getUniqueNodeContextPtr(const DAGNode *node) {
+  auto ctxIt = inputCtxs_.find(node);
+
+  if (ctxIt == inputCtxs_.end()) {
+    assert(false && "Input context not found but should exist!");
+  }
+
+  return std::move(ctxIt->second);
+}
+
+void ExecutionState::incrementInflightNodes(unsigned increment) {
+  inflightNodes_ += increment;
+}
+
+bool ExecutionState::decrementInflightNodes(unsigned decrement) {
+  // fetch_sub must be used here so that the function returns true to only one
+  // caller.
+  unsigned previousValue = inflightNodes_.fetch_sub(decrement);
+
+  // The decrement should never be more than the value of the counter at the
+  // time of decrement.
+  if (previousValue < decrement) {
+    assert(false && "More decrements than increments to inflight nodes!");
+  }
+
+  // Return true when the counter hits zero.
+  return (previousValue == decrement);
+}
+
+bool ExecutionState::incrementNodeParentsDone(const DAGNode *node,
+                                              unsigned increment) {
+  // Get the parents done counter for the node. It should have
+  // been created in the constructor.
+  auto it = nodeParentsDone_.find(node);
+
+  if (it == nodeParentsDone_.end()) {
+    assert(false && "Node parents done counter should exist but not found!");
+  }
+
+  // fetch_add must be used here so that the function returns true to only
+  // one caller.
+  unsigned numParents = (node->parents).size();
+  unsigned previousValue = (it->second).fetch_add(increment);
+  unsigned newValue = previousValue + increment;
+
+  // The new value of the counter cannot exceed the number of parents that
+  // the node has.
+  if (newValue > numParents) {
+    assert(false && "Node parents done counter incremented beyond limit!");
+  }
+
+  // Return true only when the counter hits the total numer of parents.
+  return (newValue == numParents);
+}
+
+void ExecutionState::insertIntoResultCtx(Placeholder *P, Tensor &&T) {
+  assert(resultCtx_ && "Execution result context should exist!");
+  std::lock_guard<std::mutex> lock(contextMtx_);
+  resultCtx_->insert(P, std::move(T));
+}
+
+std::unique_ptr<Context> ExecutionState::getUniqueResultContextPtr() {
+  assert(resultCtx_ && "Execution result context should exist!");
+  return std::move(resultCtx_);
```

Comment:
You need to take the context lock here, because `insertIntoResultCtx` is using that to determine if it is safe to insert into it.

Reply:
These two functions can't be concurrent due to the reasons mentioned above, but these are famous last works. I will still lock here so that I can add an assert before the `Context` is dereferenced in `insertNodeIntoCtx` to make that situation easier to debug if it happens.